examples/pipeline_tacotron2/train.py

```diff
@@ -0,0 +1,485 @@
+            if rank == 0 and writer:
+                global_iters = epoch * len(train_loader)
+                writer.add_scalar("trn/mel_loss", losses[0], global_iters)
+                writer.add_scalar("trn/mel_postnet_loss", losses[1], global_iters)
+                writer.add_scalar("trn/gate_loss", losses[2], global_iters)
+
+            trn_loss += loss * len(batch[0])
+            counts += len(batch[0])
+
+        trn_loss = trn_loss / counts
+
+        trn_loss = reduce_tensor(trn_loss, world_size)
+        if rank == 0:
+            logger.info(f"[Epoch: {epoch}] time: {time()-start}; trn_loss: {trn_loss}")
+            if writer:
+                writer.add_scalar("trn_loss", trn_loss, epoch)
+
+        if ((epoch + 1) % args.validate_and_checkpoint_freq == 0) or (epoch == args.epochs - 1):
+
+            val_start_time = time()
+            model.eval()
+
+            val_loss, counts = 0, 0
+            iterator = tqdm(enumerate(val_loader), desc=f"[Rank: {rank}; Epoch: {epoch}; Eval]", total=len(val_loader))
+
+            with torch.no_grad():
+                for val_batch_idx, val_batch in iterator:
+                    val_loss = val_loss + validation_step(model, val_batch, val_batch_idx)[0] * len(val_batch[0])
+                    counts = counts + len(val_batch[0])
+                val_loss = val_loss / counts
+
+            val_loss = reduce_tensor(val_loss, world_size)
+            if rank == 0 and writer:
+                writer.add_scalar("val_loss", val_loss, epoch)
+                log_additional_info(writer, model, val_loader, epoch)
+
+            if rank == 0:
+                is_best = val_loss < best_loss
+                best_loss = min(val_loss, best_loss)
+                logger.info(f"[Rank: {rank}, Epoch: {epoch}; Eval] time: {time()-val_start_time}; val_loss: {val_loss}")
+                logger.info(f"[Epoch: {epoch}] Saving checkpoint to {args.checkpoint_path}")
+                save_checkpoint(
+                    {
+                        "epoch": epoch + 1,
+                        "state_dict": model.state_dict(),
+                        "best_loss": best_loss,
+                        "optimizer": optimizer.state_dict(),
+                    },
+                    is_best,
+                    args.checkpoint_path,
+                )
+
+    dist.destroy_process_group()
+
+
+def main(args):
+    logger.info("Start time: {}".format(str(datetime.now())))
+
+    torch.manual_seed(0)
+    random.seed(0)
+
+    os.environ['MASTER_ADDR'] = 'localhost'
+    os.environ['MASTER_PORT'] = '17778'
+
+    device_counts = torch.cuda.device_count()
+
+    logger.info(f"# available GPUs: {device_counts}")
+
+    # download dataset is not already downloaded
+    if args.dataset == 'ljspeech':
+        if not os.path.exists(os.path.join(args.dataset_path, 'LJSpeech-1.1')):
+            from torchaudio.datasets import LJSPEECH
+            LJSPEECH(root=args.dataset_path, download=True)
+
+    if device_counts == 1:
```

Comment:
Organizing the main workflow into layers of operation will make the script more readable, for example;

Just an illustration

```python

def main():
    model = _init_modules()
    dataset = _init_dataset()
    with _init_distributed():  # <- initialize / finalize distributed training config.
        _main_train_loop()  # <- this helper is agnostic to distributed training config
```

You can look into source separation example for the actual implementation of modularization of training steps.